mlflow/utils/input_example_and_signature_utils.py

```diff
@@ -0,0 +1,76 @@
+from mlflow.models import infer_signature
```

Comment:
I like the idea of trying to DRY out the parsing of input examples / signatures; however, I think this implementation is pretty specific to the XGBoost / LightGBM use cases with the use of `_InputExampleInfo`. I also found the outputs of the functions, which are tuples of input examples, signatures, and warning messages to be somewhat difficult to reason about.

Perhaps we could instead structure this as a single `get_input_example_and_signature()` high-order function with two function arguments - `get_signature()` and `get_input_example()` - and two conf arguments `warn_input_example`, `warn_signature`.

This function could then call `get_signature` / `get_input_example` and log the appropriate warning messages.

Finally, can we place this logic in `mlflow.utils.autologging_utils` instead of defining a new module?

Reply:
I think we can reuse this in sklearn by just creating an object `_InputExampleInfo(input_example, None)`

I'm not sure what you mean by the function arguments, will these get_signature and get_input_example be defined in xgboost, lightgbm, then simply passed to this util?

We could move the logging inside the method, but I did it this way to avoid having functions with side effects. What do you think?

Yes, I will move it to the `autologging_utils` module

Reply:
Yep! We'd have each module define a function for fetching the input example and the signature (with the signature taking the input example as an argument). This gives the modules complete flexibility over how to compute the signature. Right now, we assume that signature is always going to be computed using the `model.predict()` followed by `infer_signature()` pattern.

I'd also rather not force all autologging modules to adopt InputExampleInfo when it's only going to be required for a subset of cases (XGBoost, LightGBM).

Finally, if we're concerned about side effects, we can always have the function take a `logger` argument and document that it's used to log warnings when input example / signature inference fails. Effectively, I think the goal of this utility should be to extract out common error handling / warning logic, not to extract out logic for computing the input example / signature.

Reply:
I honestly think we need to keep this InputExampleInfo, since when we collect an input example we must record the error message since get_signature needs to print this message as part of the error given to the user (e.g. `unable to infer model signature due to an error encountered when sampling a few rows of data: NoneType has no attribute slice()`)

Thus I propose the library, whether xgboost, sklearn, or lightgbm, must construct an InputExampleInfo because the error is needed at a later stage. This is passed to the helper fn, where the helper fn will check it for existence/errors and if correct then will pass it to get_signature where the signature will be computed from the InputExampleInfo. Then the helper fn will also log the messages for the user.